server/verify.go

```diff
@@ -372,3 +369,56 @@ func VerifyGceTechnology(attestation *pb.Attestation, tech pb.GCEConfidentialTec
 	}
 	return fmt.Errorf("unknown GCEConfidentialTechnology: %v", tech)
 }
+
+// parseMachineStateFromTPM is a wrapper function around `parsePCClientEventLog` method to:
+// 1. parse partial machine state from TPM TCG event logs.
+// 2. verify GceTechnology since the GCE Technology event is directly related to the TPM.
+// 3. populate the machineState TeeAttestatation field with the verified TDX/SNP attestation data.
+func parseMachineStateFromTPM(attestation *pb.Attestation, pcrs *tpmpb.PCRs, opts VerifyOpts) (*pb.MachineState, error) {
+	ms, err := parsePCClientEventLog(attestation.GetEventLog(), pcrs, opts.Loader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate the PCClient event log: %w", err)
+	}
+
+	tech := ms.GetPlatform().Technology
+	if err := verifyGceTechnology(attestation, tech, &opts); err != nil {
+		return nil, fmt.Errorf("failed to verify memory encryption technology: %w", err)
+	}
+
+	state, err := parseTEEAttestation(attestation, tech)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse machineState from TEE attestation: %w", err)
+	}
+	ms.TeeAttestation = state.TeeAttestation
+	return ms, nil
+}
+
+// parseTEEAttestation parses a machineState from TeeAttestation.
+// For now it simply populates the machineState TeeAttestation field with the verified TDX/SNP data.
+// In long term, it should parse a full machineState from TeeAttestation.
+func parseTEEAttestation(attestation *pb.Attestation, tech pb.GCEConfidentialTechnology) (*pb.MachineState, error) {
+	switch tech {
+	case pb.GCEConfidentialTechnology_AMD_SEV_SNP:
```

Comment:
I may be missing something: don't we already do these checks in verifyGceTechnology?? Can we  have that return pb.MachineState or just remove the call to parseTEEAttestation and do
`ms.TeeAttestation = attestation.TeeAttestation`?

Reply:
`ms.TeeAttestation = attestation.TeeAttestation` will not work because they're two different proto types.
`parseTEEAttestation` doesn't do any checks on attestation. This function, in short term, just populates the TeeAttestation field using a deep copy of the verified attestation data, but in long run, we'll want to get a full machineState from TDX/SNP attestation. The intent of having this function is for future-proof purpose.